### PR TITLE
feat(ci): Run Sentry integration tests

### PR DESCRIPTION
**Pull request by armenzg**
When we make changes to Relay, we should test against latest Sentry. Up until now, we would have to merge the change to Relay's `master` branch and then open a PR on Sentry to see if anything would break.

This change ensures that every PR and `master` change is tested against the latest Sentry.

#skip-changelog

### CODE REVIEW

File: .github/actions/setup-sentry/action.yml
Comment (armenzg) [inline .github/actions/setup-sentry/action.yml]: @billyvg Changing to the right directory is the main reason that I forked the action.

File: .github/workflows/ci.yml
Comment (armenzg) [inline .github/workflows/ci.yml:277]: I would have not needed to fork the action unless I was able to change directories.
I believe `setup-sentry/actions.yml` could be converted into a script and remove the need for an action.
In general, it would also help decouple from GH actions.

I think we could avoid the Docker test image work if we use this approach instead:
image

File: .github/actions/setup-sentry/action.yml
Comment (armenzg) [inline .github/actions/setup-sentry/action.yml]: I forked the action and tried to minimize the number of changes, thus, there's code that is not necessary for Relay integration tests.

File: .github/workflows/ci.yml
Comment (billyvg) [inline .github/workflows/ci.yml]:
```suggestion
      # Checkout Sentry and run integration tests against latest Relay
```

File: .github/actions/setup-sentry/action.yml
Comment (billyvg) [reply .github/actions/setup-sentry/action.yml]: Does this (working directory) persist across the rest of the steps?

File: .github/workflows/ci.yml
Comment (billyvg) [reply .github/workflows/ci.yml:277]: I think this is all CI-specific setup, so I think it's fine that it is coupled to GHA. I think *if* these actions were allowed to call anything else besides shell commands (e.g. `uses:` for other actions), actions would be better. *if* being key here :(

I was hoping that having Docker test images would cut down on the setup time, but I think you're right and we should proceed without it.

File: .github/workflows/ci.yml
Comment (armenzg) [inline .github/workflows/ci.yml]: This is only necessary until we can land https://github.com/getsentry/sentry/pull/27510

Comment (armenzg) [reply .github/workflows/ci.yml]: Do not merge until that PR lands so I can remove this line.

**Comment by armenzg** [comment]
I've tried breaking things someway, however, it's not easy to find a change that would do the trick.

If someone would like to find one for me, it would be great.

File: .github/workflows/ci.yml
Comment (billyvg) [inline .github/workflows/ci.yml]: Is this debugging?

Comment (armenzg) [reply .github/workflows/ci.yml]: Yes. I have a push in my commit history to remove it.

**Comment by untitaker** [comment]
>If someone would like to find one for me, it would be great.

giving this a shot

**Comment by armenzg** [comment]
> > If someone would like to find one for me, it would be great.
> 
> giving this a shot

Thank you!

The changes you pushed made few tests fail in [Integration Tests](https://github.com/getsentry/relay/pull/1034/checks?check_run_id=3106875262):
> FAILED tests/integration/test_store.py::test_processing[default] - assert Non...
> FAILED tests/integration/test_store.py::test_processing[transaction] - assert...
> FAILED tests/integration/test_unreal.py::test_unreal_minidump_with_processing
> FAILED tests/integration/test_unreal.py::test_unreal_apple_crash_with_processing
> FAILED tests/integration/test_unreal.py::test_unreal_minidump_with_config_and_processing

but not in `Sentry-Relay integration tests`.

I'm not looking for a change that would break Relay but one that would pass *yet* the integration tests would fail.

**Comment by armenzg** [comment]
As per my conversation with Markus, this is still testing against Relay's nightly image.
See this file for details:
https://github.com/getsentry/sentry/blob/master/src/sentry/utils/pytest/relay.py

**Comment by armenzg** [comment]
@untitaker Building the Relay Docker image is super slow. I've tried [three various Docker caching strategies](https://github.com/getsentry/relay/blob/7eaef3bc4830199448d98ba5c27e65573536df92/.github/workflows/ci.yml#L235-L284), however, the step that takes 20 minutes is `cd relay && cargo build --release --locked --features ssl,processing --target=x86_64-unknown-linux-gnu`:
https://github.com/getsentry/relay/runs/3127050302?check_suite_focus=true#step:3:705

On GRC, it seems to take 8 minutes:
https://console.cloud.google.com/cloud-build/builds/7252f715-4bbd-4b46-8176-b032e7a43d90;step=4;tab=detail?project=sentryio

**Comment by armenzg** [comment]
@untitaker @BYK @jan-auer

Hi,
This PR can run integration tests from Sentry against the latest Relay. Now, this can take over 20 minutes and most of it is building the Relay Docker image. I did some optimizations (from 20 minutes to 17 minutes) but I can't get the `relay-builder` stage cached.

I have done some optimizations to the Docker process by building with BUILDKIT_INLINE_CACHE=1 (which allows using --cache-from ) and publishes it to GH's registry (since the BUILDKIT_INLINE_CACHE is not in use for Cloud builds) and make's Sentry integration tests run against this image thanks to the variable `RELAY_TEST_IMAGE` (see [sentry PR](https://github.com/getsentry/sentry/pull/27596)) for details).

The slowest step is [make build-linux-release TARGET=x86_64-unknown-linux-gnu RELAY_FEATURES=ssl,processing](https://github.com/getsentry/relay/blob/fba39474ec9d5d1ce941797d9927d71087053b6d/Dockerfile#L38-L39) which takes 14 minutes. I believe it *always* executes because of our multi-stage Docker builds (see the word CACHE until we start the 2nd stage - `relay-builder`):
image

To my surprise, stage 3 is cached:
image

Any thoughts on we can we achieve the `make build-linus-release` to be cached?
Here's the steps that I want cached https://github.com/getsentry/relay/blob/fba39474ec9d5d1ce941797d9927d71087053b6d/Dockerfile#L29-L39

Should we take this PR even if it's slow so we can prevent potential deployment issues?
Or should we directly work on [relay.py](https://github.com/getsentry/sentry/blob/master/src/sentry/utils/pytest/relay.py) to support non-Docker testing? (that file currently starts/stops the Relay *nightly* container for testing).

I think I should open a new PR for BUILDKIT_INLINE_CACHE for the cloudbuild builds regardless of what we do in this PR.
I could then use `--cache-from` directly from GC's registry as the starting image.

**Comment by armenzg** [comment]
It seems that tests itself are not quite executing. I have probably regressed something with all these optimizations.
I will look more into it next week.

```
________________________ ERROR at setup of test_simple _________________________
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/api/client.py", line 261, in _raise_for_status
    response.raise_for_status()
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http+docker://localhost/v1.35/containers/d7a02fea32bc829c1be6fb169c494ea171792bc18cc2b43a0707f0ed36a029b2/start

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/relay/relay/sentry/src/sentry/utils/pytest/relay.py", line 126, in relay_server
    docker_client.containers.run(**options)
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/models/containers.py", line 791, in run
    container.start()
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/models/containers.py", line 392, in start
    return self.client.api.start(self.id, **kwargs)
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/api/container.py", line 1091, in start
    self._raise_for_status(res)
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/api/client.py", line 263, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/opt/hostedtoolcache/Python/3.6.14/x64/lib/python3.6/site-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.APIError: 400 Client Error: Bad Request ("OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "run": executable file not found in $PATH: unknown")
```

**Comment by untitaker** [comment]
>Any thoughts on we can we achieve the make build-linus-release to be cached?

I have no expertise on how docker caching works on GCB or GHA. I even believe that somebody outside of our team set this up for us, the entirety of it.

I am fine with merging as-is. @jan-auer wdyt?

**Comment by BYK** [comment]
>I have no expertise on how docker caching works on GCB or GHA. I even believe that somebody outside of our team set this up for us, the entirety of it.

I think it was me 🙂 Talking to @armenzg privately about this.

**Comment by armenzg** [comment]
I've changed the strategy for what I hope to be the fastest solution we have available (instead of running tests against non-Docker Relay).

Yesterday, I tried making the GCB trigger a workflow via `workdlow_dispatch` (see #1047), however, it did not update the commit status.

Today, I've made the GCB push the image for the commit ASAP (about 11 minutes after starting). Meanwhile, the integration tests poll for the GC image until it finds it, thus, we can finish the integration tests in 16 minutes (1/3 faster) rather than 24 minutes (Building the Docker image in GHA).

File: .github/workflows/ci.yml
Comment (armenzg) [inline .github/workflows/ci.yml]: We will remove this when https://github.com/getsentry/sentry/pull/27596 merges.

**Review by untitaker — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (billyvg) [inline .github/workflows/ci.yml]: I think this should work

```
SHA=${{ github.pull_request.head.sha || github.sha }}
```

**Review by billyvg — APPROVED** [review]
LGTM after removing the temp checkout ref

**Review by BYK — APPROVED** [review]
LGTM

I think we can benefit from having a sentry-test image in the future to speed up certain things.

**Pull request merged**
